pkg/types/cose/v0.0.1/entry.go

```diff
@@ -0,0 +1,350 @@
+	if err := json.Unmarshal(b, &stmt); err != nil {
+		return nil, err
+	}
+
+	return &stmt, nil
+}
+
+func formatKey(b []byte) string {
+	h := sha256.Sum256(b)
+	hash := hex.EncodeToString(h[:])
+	return strings.ToLower(fmt.Sprintf("%s:%s", models.CoseV001SchemaDataPayloadHashAlgorithmSha256, hash))
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	it, ok := pe.(*models.Cose)
+	if !ok {
+		return errors.New("cannot unmarshal non Cose v0.0.1 type")
+	}
+
+	var err error
+	if err := types.DecodeEntry(it.Spec, &v.CoseObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.CoseObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	v.keyObj, err = x509.NewPublicKey(bytes.NewReader(*v.CoseObj.PublicKey))
+	if err != nil {
+		return err
+	}
+
+	// Store the envelope hash.
+	// The CoseObj.Message is only populated during entry creation.
+	// When marshalling from the database (retrieval) the envelope
+	// hash must be decoded fromt he stored hex string.
+	// The envelope hash is used to create the attestation key during
+	// retrieval of a record.
+	if len(v.CoseObj.Message) == 0 {
+		b, err := hex.DecodeString(*v.CoseObj.Data.EnvelopeHash.Value)
+		if err != nil {
+			return err
+		}
+		v.envelopeHash = b
+	} else {
+		h := sha256.Sum256(v.CoseObj.Message)
+		v.envelopeHash = h[:]
+	}
+
+	return v.validate()
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	if v.keyObj == nil {
+		return nil, errors.New("cannot canonicalze empty key")
+	}
+	pk, err := v.keyObj.CanonicalValue()
+	if err != nil {
+		return nil, err
+	}
+	pkb := strfmt.Base64(pk)
+
+	h := sha256.Sum256([]byte(v.sign1Msg.Payload))
+
+	canonicalEntry := models.CoseV001Schema{
+		PublicKey: &pkb,
+		Data: &models.CoseV001SchemaData{
+			PayloadHash: &models.CoseV001SchemaDataPayloadHash{
+				Algorithm: swag.String(models.CoseV001SchemaDataPayloadHashAlgorithmSha256),
+				Value:     swag.String(hex.EncodeToString(h[:])),
+			},
+			EnvelopeHash: &models.CoseV001SchemaDataEnvelopeHash{
+				Algorithm: swag.String(models.CoseV001SchemaDataEnvelopeHashAlgorithmSha256),
+				Value:     swag.String(hex.EncodeToString(v.envelopeHash)),
+			},
+		},
+	}
+
+	itObj := models.Cose{}
+	itObj.APIVersion = swag.String(APIVERSION)
+	itObj.Spec = &canonicalEntry
+
+	return json.Marshal(&itObj)
+}
+
+// validate performs cross-field validation for fields in object
+func (v *V001Entry) validate() error {
+	// This also gets called in the CLI, where we won't have this data
+	// or during record retrieval (message is the raw COSE object) which
+	// is only stored as an attestation.
+	if len(v.CoseObj.Message) == 0 {
+		return nil
+	}
+
+	alg, pk, err := getPublicKey(v.keyObj)
+	if err != nil {
+		return nil
+	}
+
+	bv, err := gocose.NewVerifier(alg, pk)
+	if err != nil {
+		return err
+	}
+	v.sign1Msg = gocose.NewSign1Message()
+	if err := v.sign1Msg.UnmarshalCBOR(v.CoseObj.Message); err != nil {
+		return err
+	}
+
```

Comment:
should we ensure unmarshaling is successful before assigning the output to `v.sign1Msg`?

Reply:
You mean like this:
```go
        coseMsg := gocose.NewSign1Message()
	if err := coseMsg.UnmarshalCBOR(v.CoseObj.Message); err != nil {
		return err
	}
        v.sign1Msg = coseMsg
```

I think that makes sense, we can even wait until the signature is verified until we assign it to `v.sign1Msg`.

Reply:
yes, ideally I wouldn't want to partially change the pointer receiver unless all validation is successful.